File: myFinance_improved_V2/lib/features/auth/data/models/user_model.dart
// lib/features/auth/data/models/user_model.dart

import '../../domain/entities/user_entity.dart';
import '../../../../core/utils/datetime_utils.dart';

/// User Model
///
/// 📦 택배 상자 - JSON 직렬화 가능한 데이터 모델
///
/// 책임:
/// - JSON ↔ Dart 객체 변환
/// - Database 컬럼명 매핑 (user_id, first_name, last_name, etc.)
/// - Entity 변환
///
/// 이 모델은 Supabase users 테이블 JSON 구조에 대한 지식을 가지고 있습니다.
class UserModel {
  final String userId;
  final String email;
  final String? firstName;
  final String? lastName;
  final String? userPhoneNumber;
  final String? profileImage;
  final String? preferredTimezone;
  final String createdAt;
  final String? updatedAt;
  final bool isDeleted;
  final bool isEmailVerified;
  final String? lastLoginAt;

  const UserModel({
    required this.userId,
    required this.email,
    this.firstName,
    this.lastName,
    this.userPhoneNumber,
    this.profileImage,
    this.preferredTimezone,
    required this.createdAt,
    this.updatedAt,
    this.isDeleted = false,
    this.isEmailVerified = false,
    this.lastLoginAt,
  });

  /// Create from Supabase JSON
  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      userId: json['user_id'] as String,
      email: json['email'] as String,
      firstName: json['first_name'] as String?,
      lastName: json['last_name'] as String?,
      userPhoneNumber: json['user_phone_number'] as String?,
      profileImage: json['profile_image'] as String?,
      preferredTimezone: json['preferred_timezone'] as String?,
      createdAt: json['created_at'] as String,
      updatedAt: json['updated_at'] as String?,
      isDeleted: json['is_deleted'] as bool? ?? false,
      isEmailVerified: json['is_email_verified'] as bool? ?? false,
      lastLoginAt: json['last_login_at'] as String?,
    );
  }

  /// Convert to Supabase JSON
  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'email': email,
      'first_name': firstName,
      'last_name': lastName,
      'user_phone_number': userPhoneNumber,
      'profile_image': profileImage,
      'preferred_timezone': preferredTimezone,
      'created_at': createdAt,
      'updated_at': updatedAt,
      'is_deleted': isDeleted,
      'is_email_verified': isEmailVerified,
      'last_login_at': lastLoginAt,
    };
  }

  /// Convert to Domain Entity
  User toEntity() {
    return User(
      id: userId,
      email: email,
      firstName: firstName,
      lastName: lastName,
      createdAt: DateTimeUtils.toLocal(createdAt),
      lastLoginAt: lastLoginAt != null ? DateTimeUtils.toLocal(lastLoginAt!) : null,
      isEmailVerified: isEmailVerified,
    );
  }

  /// Create from Domain Entity
  factory UserModel.fromEntity(User entity) {
    return UserModel(
      userId: entity.id,
      email: entity.email,
      firstName: entity.firstName,
      lastName: entity.lastName,
      createdAt: DateTimeUtils.toUtc(entity.createdAt),
      lastLoginAt: entity.lastLoginAt != null ? DateTimeUtils.toUtc(entity.lastLoginAt!) : null,
      isEmailVerified: entity.isEmailVerified,
    );
  }

  /// Create insert map for Supabase (user profile creation)
  ///
  /// Used when creating a new user profile in the users table.
  /// The user_id should match the auth.users.id from Supabase Auth.
  Map<String, dynamic> toInsertMap() {
    final map = <String, dynamic>{
      'user_id': userId,
      'email': email,
      'created_at': createdAt,
      'updated_at': DateTimeUtils.nowUtc(),
    };

    if (firstName != null) map['first_name'] = firstName;
    if (lastName != null) map['last_name'] = lastName;
    if (userPhoneNumber != null) map['user_phone_number'] = userPhoneNumber;
    if (profileImage != null) map['profile_image'] = profileImage;
    if (preferredTimezone != null) map['preferred_timezone'] = preferredTimezone;

    return map;
  }

  /// Create update map for Supabase
  ///
  /// Only includes fields that can be updated.
  /// Does not include user_id, email, or created_at.
  Map<String, dynamic> toUpdateMap() {
    final map = <String, dynamic>{
      'updated_at': DateTimeUtils.nowUtc(),
    };

    if (firstName != null) map['first_name'] = firstName;
    if (lastName != null) map['last_name'] = lastName;
    if (userPhoneNumber != null) map['user_phone_number'] = userPhoneNumber;
    if (profileImage != null) map['profile_image'] = profileImage;
    if (preferredTimezone != null) map['preferred_timezone'] = preferredTimezone;

    return map;
  }

  /// Copy with updated fields
  UserModel copyWith({
    String? userId,
    String? email,
    String? firstName,
    String? lastName,
    String? userPhoneNumber,
    String? profileImage,
    String? preferredTimezone,
    String? createdAt,
    String? updatedAt,
    bool? isDeleted,
    bool? isEmailVerified,
    String? lastLoginAt,
  }) {
    return UserModel(
      userId: userId ?? this.userId,
      email: email ?? this.email,
      firstName: firstName ?? this.firstName,
      lastName: lastName ?? this.lastName,
      userPhoneNumber: userPhoneNumber ?? this.userPhoneNumber,
      profileImage: profileImage ?? this.profileImage,
      preferredTimezone: preferredTimezone ?? this.preferredTimezone,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isDeleted: isDeleted ?? this.isDeleted,
      isEmailVerified: isEmailVerified ?? this.isEmailVerified,
      lastLoginAt: lastLoginAt ?? this.lastLoginAt,
    );
  }
}

File: website/src/features/auth/data/models/UserModel.ts
/**
 * User Model (DTO)
 * Data layer - Data Transfer Object for User
 */

import { User as SupabaseUser } from '@supabase/supabase-js';
import { User } from '../../domain/entities/User';

export class UserModel {
  /**
   * Convert Supabase User to Domain User Entity
   */
  static fromSupabase(supabaseUser: SupabaseUser): User {
    return User.create({
      id: supabaseUser.id,
      email: supabaseUser.email!,
      firstName: supabaseUser.user_metadata?.first_name || null,
      lastName: supabaseUser.user_metadata?.last_name || null,
      emailConfirmed: supabaseUser.email_confirmed_at !== null,
      createdAt: supabaseUser.created_at,
      lastSignIn: supabaseUser.last_sign_in_at,
      metadata: supabaseUser.user_metadata || {},
      userPhoneNumber: supabaseUser.user_metadata?.user_phone_number || null,
      profileImage: supabaseUser.user_metadata?.profile_image || null,
      preferredTimezone: supabaseUser.user_metadata?.preferred_timezone || 'Asia/Ho_Chi_Minh',
      isDeleted: false,
      updatedAt: supabaseUser.updated_at || null,
    });
  }

  /**
   * Convert users table row to Domain User Entity
   */
  static fromUsersTable(row: {
    user_id: string;
    email: string;
    first_name: string | null;
    last_name: string | null;
    email_confirmed_at?: string | null;
    created_at: string;
    last_sign_in_at?: string | null;
    user_phone_number?: string | null;
    profile_image?: string | null;
    preferred_timezone?: string;
    is_deleted?: boolean;
    updated_at?: string | null;
  }): User {
    return User.create({
      id: row.user_id,
      email: row.email,
      firstName: row.first_name,
      lastName: row.last_name,
      emailConfirmed: row.email_confirmed_at !== null,
      createdAt: row.created_at,
      lastSignIn: row.last_sign_in_at || null,
      metadata: {},
      userPhoneNumber: row.user_phone_number || null,
      profileImage: row.profile_image || null,
      preferredTimezone: row.preferred_timezone || 'Asia/Ho_Chi_Minh',
      isDeleted: row.is_deleted || false,
      updatedAt: row.updated_at || null,
    });
  }

  /**
   * Convert Domain User Entity to users table insert format
   * Following Flutter implementation: user_model.dart toInsertMap()
   */
  static toInsertMap(user: User): Record<string, any> {
    const map: Record<string, any> = {
      user_id: user.id,
      email: user.email,
      created_at: user.createdAt.toISOString(),
      updated_at: new Date().toISOString(),
    };

    if (user.firstName) map.first_name = user.firstName;
    if (user.lastName) map.last_name = user.lastName;
    if (user.userPhoneNumber) map.user_phone_number = user.userPhoneNumber;
    if (user.profileImage) map.profile_image = user.profileImage;
    if (user.preferredTimezone) map.preferred_timezone = user.preferredTimezone;

    return map;
  }

  /**
   * Convert Domain User Entity to JSON
   */
  static toJson(user: User): Record<string, any> {
    return {
      id: user.id,
      email: user.email,
      first_name: user.firstName,
      last_name: user.lastName,
      email_confirmed: user.emailConfirmed,
      created_at: user.createdAt.toISOString(),
      last_sign_in_at: user.lastSignIn?.toISOString() || null,
      metadata: user.metadata,
      user_phone_number: user.userPhoneNumber,
      profile_image: user.profileImage,
      preferred_timezone: user.preferredTimezone,
      is_deleted: user.isDeleted,
      updated_at: user.updatedAt?.toISOString() || null,
    };
  }

  /**
   * Convert Domain User Entity to update map for Supabase
   * Following Flutter implementation: user_model.dart toUpdateMap()
   * Only includes fields that can be updated
   */
  static toUpdateMap(user: User): Record<string, any> {
    const map: Record<string, any> = {
      updated_at: new Date().toISOString(),
    };

    if (user.firstName) map.first_name = user.firstName;
    if (user.lastName) map.last_name = user.lastName;
    if (user.userPhoneNumber) map.user_phone_number = user.userPhoneNumber;
    if (user.profileImage) map.profile_image = user.profileImage;
    if (user.preferredTimezone) map.preferred_timezone = user.preferredTimezone;

    return map;
  }
}
